### ISSUE TITLE
Modifying AST tree

### ISSUE THREAD

**Issue by djn3m0** [austinbyers/esprima-ast-visitor#1]
Hello,

Im visiting the AST tree generated by acorn, it works perfectly but i want to write down the generated tree that im visiting to a file to use esgenerate to make new .js file.

thanks

thanks

**Comment by austinbyers** [comment]
So you want to be able to save the modified tree back out to a file? That shouldn't be too difficult, I'll see if I can add that soon

**Comment by djn3m0** [comment]
Yes. 
Now im visiting the ast tree like:
```
for node in visitor.traverse(json.loads(esprima_ast_string)):
    print(node['type'])
```
and it is good and powerful. I goes over each node etc...
It would be great to have same thing but for save/reconstructing the tree. 

For example: Traverse over each node of the tree and make a modification in a MemberExpression or any other node, and apply that modification to the tree.
```
 {
            "type": "ExpressionStatement",
            "start": 66,
            "end": 88,
            "expression": {
              "type": "CallExpression",
              "start": 66,
              "end": 87,
              "callee": {
                "type": "Identifier",
                "start": 66,
                "end": 71,
                "name": "alert"
              },
              "arguments": [
                {
                  "type": "Literal",
                  "start": 72,
                  "end": 86,
                  "value": "foo function",
                  "raw": "'foo function'"
                }
              ]
            }
          }
```

What im getting from visiting that node is:

```
ExpressionStatement
CallExpression
Identifier
Literal
```
Which if i save the node to a file i will save ExpressionStatement, then CallExpression, then Identifier, then Literal, all seperately.
I tried to save node directly to a file but as you know nodes are not in the order so the output tree was so broken! How can I put all those nodes together to create the new tree like the original one?

**Comment by austinbyers** [comment]
Right now, the code only does a [pre-order traversal](https://en.wikipedia.org/wiki/Tree_traversal#Pre-order); it doesn't actually store the tree structure. So right now there isn't a good way to modify and write back.

However, I'm working on a rewrite that puts everything into Python objects that can be traversed, modified, saved, etc.

Are there any features that you'd like to have?

**Comment by austinbyers** [comment]
@djn3m0 thanks for the suggestion; I added some more tests and released v1.0 with this fix. Let me know if it ends up being what you need!
